### ISSUE TITLE
Execute a callback when tag is used

### ISSUE THREAD

**Issue by RikSomers** [s9e/TextFormatter#53]
Hi,

I'm working on a website where I have the requirement to execute a callback when a given tag is being used. For example, I want to send a message to a user (via a text message) that he has been mentioned in a thread. I would love to be able to simply register a callback to, for example, the quote-tag if this is possible, that checks if the referenced "author" variable is some person that is registered to my site.

```
[quote=Rik]...[/quote]

Nice post!
```

Would then need to call something like this (just a simplified pseudocode example):
```
Parser::tag('quote')->callback(function($quote) {
        $user = User::find($quote->author);
        $user->notify('You have been quoted');
    });
```

What would be the best way to do this with your package?
Is this something that is (natively) supported, or if not, is this something you would be willing to implement?

Thanks in advance!

**Comment by JoshyPHP** [comment]
[Tag filters](http://s9etextformatter.readthedocs.io/Filters/Tag_filters/) are executed during parsing and could be used for that, but that's not what I'd recommend. Sample code for illustration:

```php
$configurator->tags['quote']->filterChain->append('myCallback');
function myCallback($tag)
{
	$user = User::find($quote->getAttribute('author'));
	$user->notify('You have been quoted');

	return true;
}
```

What I recommend is to parse the text normally then extract the quote authors from the XML using DOM.

```php
$dom = new DOMDocument;
$dom->loadXML($xml);
foreach ($dom->getElementsByTagName('QUOTE') as $quote)
{
	$user = User::find($quote->getAttribute('author'));
	$user->notify('You have been quoted');
}
```

**Comment by RikSomers** [comment]
Alright. Tag Filters seem to be the cleanest way of doing real "callbacks" during parsing. I don't like iterating over each element again after parsing.

For example, another use-case I have is not parsing a certain tag when a user does not have permissions to use that tag.

It would seem to me the cleanest way for such a tag would be using the filterChain, for example (again in simplified pseudocode):
```
$configurator->tags['mytag']->filterChain->append('myCallback');
function myCallback($tag)
{
	return CurrentLoggedInUser::canUseTag('mytag'); // returns true or false
}
```

It seems to me that with that use-case, your second approach does not work.

It would seem to me that a native solution to add callbacks to beforeParse / afterParse of a given tag would be a better approach.

**Comment by JoshyPHP** [comment]
It's generally more efficient to run that kind of routines on the XML. You can deduplicate the values more easily, and you can retrieve the user records in bulk rather than individually.

You can toggle tags before parsing if you know which ones should be disabled beforehand.
http://s9etextformatter.readthedocs.io/Getting_started/Runtime_configuration/#toggle-a-tag

If you want a custom callback to (in)validate a tag then yes, tag filters are the way to do it.

**Comment by RikSomers** [comment]
Ah right, I knew I saw something about that in the docs. That might indeed be a better solution for the second use-case I provided.

I get what you're saying with the efficiency of callbacks, and I agree it would probably be better performance-wise when, for example, working with lots of tags like quotes. I guess it's a tradeoff between code simplicity/cleanliness (which is ofcourse subjective) and performance by iterating over the XML after parsing.

Either way, I can't think of any other use-cases at the moment that would require the type of callback routine I proposed in previous comments, and most of them can be fixed by using the XML-based iterations or Tag filters.

Thanks for the quick responses!
